src/settings/RepositoryTableModel.cpp

```diff
@@ -0,0 +1,151 @@
+#include "RepositoryTableModel.h"
+#include <QSet>
+
+RepositoryTableModel::RepositoryTableModel(QObject * parent)
+: QAbstractTableModel(parent) {
+    load({});
+}
+
+QList<QStringList> RepositoryTableModel::dump() const {
+    QList<QStringList> out;
+    for (auto &&repo : repositories_) {
+        if (repo.isDefault) continue;
+        out << QStringList{repo.name, repo.url};
+    }
+    return out;
+}
+
+void RepositoryTableModel::load(QList<QStringList> data) {
+    beginRemoveRows(QModelIndex(), 0, repositories_.size());
+    repositories_.clear();
+    urls_.clear();
+    endRemoveRows();
+
+    repositories_ << Repository{kDefaultRepositoryName,kDefaultRepositoryURL, true};
+    urls_.insert(kDefaultRepositoryURL);
+    
+    for (auto &&pair : data) {
+        // pair: [name, ..., url] where ... is currently [].
+
+        if (urls_.contains(pair.back()))
+            continue;
+
+        repositories_ << Repository{pair.first(), pair.back(), false};
+        urls_.insert(pair.back());
+    }
+        
+    // TODO: I'm lying here, it has already happened. Will Qt be okay with that?
+    beginInsertRows(QModelIndex(), 0, repositories_.size());
+    endInsertRows();
+}
+
+bool RepositoryTableModel::canRemove(QModelIndex index) const {
+    return !repositories_.at(index.row()).isDefault;
+}
+
+void RepositoryTableModel::insert(QString name, QString url) {
+    if (urls_.contains(url)) {
+        // TODO: update name instead?
```

Comment:
The user can just delete and add again, too much work otherwise.

Reply:
I agree. Maybe at some point I'll make the table properly editable so you can double-click a field and change the names or urls that way.